gtsfm/evaluation/metrics.py

```diff
@@ -0,0 +1,287 @@
+        BAR = 1  # For scalars
+        BOX = 2  # For 1D distributions
+        HISTOGRAM = 3  # For 1D distributions
+
+    def _get_plot_types_for_dim(self, dim) -> List[PlotType]:
+        if dim == 0:
+            return [self.PlotType.BAR]
+        if dim == 1:
+            return [self.PlotType.BOX, self.PlotType.HISTOGRAM]
+        return []
+
+    def _get_distribution_histogram(self, data: np.ndarray) -> Dict[str, Union[float, int]]:
+        """Returns the histogram of data as a dictionary.
+
+        If the data is float, the keys of the dictionary are interval buckets, and if the data is int, the keys are also int.
+
+        Args:
+            data: 1D array of all values of the metric
+
+        Returns:
+            Histogram of data as a dict from bucket to count.
+        """
+        if data.size == 0:
+            print("Requested histogram for empty data metric, returning None.")
+            return None
+        if isinstance(data.tolist()[0], int):
+            # One bin for each integer
+            bins = int(np.max(data) - np.min(data) + 1)
+            discrete = True
+        else:
+            bins = 10
+            discrete = False
+        count, bins = np.histogram(data, bins=bins)
+        count = count.tolist()
+        bins = bins.tolist()
+        bins_lower = bins[:-1]
+        bins_upper = bins[1:]
+
+        histogram = {}
+        for i in range(len(count)):
+            if discrete:
+                key = str(int(bins_lower[i]))
+            else:
+                key = "%.2f-%.2f" % (bins_lower[i], bins_upper[i])
+            histogram[key] = count[i]
+        return histogram
+
+    def _create_summary(self, data: np.ndarray) -> Dict[str, Any]:
+        """Creates a summary of the given data.
+
+        This is useful for analysis as data can be very large. The summary is a dict contains the following fields:
+            - Min, max, median of data
+            - Mean and std dev of data
+            - Either quartiles or histogram of the data depending on plot_type of this metric.
+
+        Args:
+            data: 1D array of all values of the metric
+
+        Returns:
+            summary as a dict that can be serialized to JSON for storage.
+        """
+        if data.ndim != 1:
+            raise ValueError("Metric must be a 1D distribution to get summary.")
+        summary = {
+            "min": np.min(data).tolist(),
+            "max": np.max(data).tolist(),
+            "median": np.median(data).tolist(),
+            "mean": np.mean(data).tolist(),
+            "stddev": np.std(data).tolist(),
+        }
+        if self._plot_type == self.PlotType.BOX:
+            summary.update({"quartiles": self._get_distribution_quartiles(data)})
+        elif self._plot_type == self.PlotType.HISTOGRAM:
+            summary.update({"histogram": self._get_distribution_histogram(data)})
+        return summary
+
+    def _get_distribution_quartiles(self, data: np.ndarray) -> Dict[int, float]:
```

Comment:
Where were you thinking we would use quartile info? Just for dumping to stdout or json?

Reply:
The quartiles are used for the box plots if raw data is not stored. Raw data can be really big for things like track lengths or reprojection errors, and should not be stored.

Reply:
It could also be used for logging if thats necessary.

Reply:
i think all of our data is still pretty small for now. Weren't you using full data to show all the plotly plots this past week?

Reply:
or i guess you were just saving this histogram bin counts a priori, and quartiles for the box plots a priori? this could be helpful to put in the module description as a high-level summary of the file's purpose, in line 2 of this file

Reply:
> i think all of our data is still pretty small for now.

No, I dont think the reprojection errors are small. data association and/or bundle adjustment metric JSON files are not human readable because there are too many lines. Also, I found some metrics (probably in bundle adjustment) where we store the mean, min, max etc of the same metric without storing the full data.

Reply:
> or i guess you were just saving this histogram bin counts a priori, and quartiles for the box plots a priori? this could be helpful to put in the module description as a high-level summary of the file's purpose, in line 2 of this file

Yeah, that is what I was doing when we dont have full data. I think should be documented in the module that does the plotting, not here.

Reply:
this would be nice as a free function with a unit test